src/main/java/com/conveyal/gtfs/model/LocationShape.java

```diff
@@ -95,6 +102,84 @@ public String toCsvRow() {
         ) + System.lineSeparator();
     }
 
+    /**
+     * Validate location geometry types. Specification: https://www.rfc-editor.org/rfc/rfc7946#section-3.1.6.
+     */
+    public static JsonNode validate(JsonNode jsonNode) throws IOException {
+        JsonNode geometryType = jsonNode.get("geometry_type");
+        String type = geometryType.asText();
+        switch (type) {
+            case "polygon":
+                return validatePolygon(jsonNode);
+            // TODO: Add other geometry types when they are supported.
+            default:
+                throw new IOException(String.format("Geometry type: %s, is not supported.", type));
+        }
+    }
+
+    /**
+     * A valid polygon must have at least four shapes and the first and last shapes must be the same. If only three
+     * shapes are provided and the first and last shapes match an error is thrown. If more than three shapes are
+     * provided and the first and last shapes do not match an additional shape matching the first is added to the end.
+     */
+    private static JsonNode validatePolygon(JsonNode jsonNode) throws IOException {
+        ArrayNode locationShapes = (ArrayNode) jsonNode.get("location_shapes");
+        Iterator<JsonNode> shapes = locationShapes.elements();
+        ObjectNode firstShape = null;
+        ObjectNode lastShape = null;
+        int count = 0;
+        while (shapes.hasNext()) {
+            JsonNode shape = shapes.next();
+            count++;
+            if (count == 1) {
+                firstShape = getLocationShape(shape);
+            } else {
+                lastShape = getLocationShape(shape);
+            }
+        }
+        if (count <= 2) {
+            throw new IOException(shapeCountErrorMessage);
+        }
+
+        boolean shapesAreTheSame = shapesMatch(firstShape, lastShape);
+        if (shapesAreTheSame && count == 3) {
+            // Polygon has been closed, but there are not enough shapes provided.
+            throw new IOException(shapeCountErrorMessage);
+        } else if (!shapesAreTheSame) {
+            // Add a new shape to the end of the array matching the first shape. Increment the last shape id by one to
+            // create a unique id for this shape.
+            int lastShapeId = lastShape.get("id").asInt();
+            firstShape.put("id", ++lastShapeId);
+            locationShapes.add(firstShape);
```

Comment:
Should we add a warning here that the shape was automatically closed?

Reply:
Good point. Updated.